### PR TITLE
Create OpenStack Server #513  - Enhanced Create Server.

### LINKED ISSUE #513: Create OpenStack Server
Existing operation has hard-coded values in REST HTTP Post API body call, so the functionality is severe limited. There are also certain inputs that are missing (regarding: key-pairs, security-groups, metadata, user-data, version, etc).

### PR DESCRIPTION
**Pull request by mihaitusa**
fixes #513 Signed-off-by: Mihail-Cristian Tusa <mihail-cristian.tusa@hpe.com>

### CODE REVIEW

**Comment by mihaitusa** [comment]
@Bonczidai  Please review it. DYT that helper ops from /utils should be made more generic and reused them in base/json?

**Comment by mihaitusa** [comment]
closes #513

File: content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl [hunk 1]
```diff
@@ -7,36 +7,86 @@
...
+#! @input network_id: optional - ID of network to connect to - Default: ''
 #! @output error_message: return_result if statusCode is not '202'
 #! @output return_code: '0' if success, '-1' otherwise
 #! @output status_code: code returned by operation
 #! @result SUCCESS: OpenStack server (instance) was successfully created
 #! @result GET_AUTHENTICATION_TOKEN_FAILURE: authentication token cannot be obtained from authentication call response
 #! @result GET_TENANT_ID_FAILURE: tenant_id corresponding to tenant_name cannot be obtained from authentication call response
-#! @result GET_AUTHENTICATION_FAILURE: authentication call fails
-#! @result CREATE_SERVER_FAILURE: OpenStack server (instance) could not be created
+#! @result ADD_NAME_FAILURE:
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl]: remove `:`

File: content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl [hunk 2]
```diff
@@ -7,36 +7,86 @@
...
+#! @result ADD_KEY_NAME_FAILURE
+#! @result ADD_SECURITY_GROUPS_FAILURE
+#! @result ADD_USER_DATA_FAILURE
+#! @result ADD_AVAILABILITY_ZONE_FAILURE
+#! @result ADD_NETWORK_FAILURE
+#! @result ADD_METADATA_FAILURE
+#! @result ADD_PERSONALITY_FAILURE
+#! @result ADD_INSTANCE_SOURCE_BOOT_FAILURE
+#! @result ADD_SERVER_OBJECT_JSON_FAILURE
+#! @result CREATE_SERVER_FAILURE: could not create OpenStack server (instance)
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl]: IMHO the failures are pretty self descriptive, so you can leave them `#! @result ADD_PERSONALITY_FAILURE` form - in that case do this for all the failures

File: content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl [hunk 3]
```diff
@@ -60,6 +107,39 @@ flow:
         required: false
     - proxy_password:
         required: false
+    - tenant_name
+    - version: '2'
+    - server_name
+    - img_ref
+    - flavor_ref: '42'
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl]: just make sure to specify the `required` property properly for all the inputs - keep in mind that `required` is not related to the default value but to the `overridable` property

when deciding on required property, ask this question: after binding can this input be null?

File: content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl [hunk 4]
```diff
@@ -75,33 +155,312 @@ flow:
...
+          - error_message
+          - return_code
+        navigate:
+          SUCCESS: validate_user_data
+          FAILURE: ADD_ADMIN_PASS_FAILURE
+
+    - validate_user_data:
+        do:
+          strings.string_equals:
+            - first_string: ${user_data}
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl]: AFAIU if the user data is not empty you want to add it to the json - but what happens if the user passes an integer as user_data?

One idea is to create is_empty(null + '' check like in Java) utility op in `strings` and use `str(user_data)` either here or in inputs section with overridable false

applies to all the similar inputs

WDYT?

File: content/io/cloudslang/cloud/openstack/utils/add_json_object.sl [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        return_code = '-1'
+
+  outputs:
+    - json_output: ${ encoded_json if return_code == '0' else '' }
+    - return_result
+    - return_code
+    - error_message: ${ return_result if return_code == '-1' else '' }
+
+  results:
+    - SUCCESS
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/utils/add_json_object.sl]: missing SUCCESS condition

File: content/io/cloudslang/cloud/openstack/utils/add_json_object.sl [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+#! @output error_message: error message if there was an error when executing, empty otherwise
+#! @result SUCCESS: successfully inserted the JSON object (return_code == '0')
+#! @result FAILURE: otherwise
+#!!#
+####################################################
+
+namespace: io.cloudslang.cloud.openstack.utils
+
+operation:
+  name: add_json_object
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/utils/add_json_object.sl]: add_entry_in_object under json namespace?

File: content/io/cloudslang/cloud/openstack/utils/update_list.sl [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        return_code = '-1'
+
+  outputs:
+    - json_output: ${ encoded_json if return_code == '0' else '' }
+    - return_result
+    - return_code
+    - error_message: ${ return_result if return_code == '-1' else '' }
+
+  results:
+    - SUCCESS
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/utils/update_list.sl]: missing condition

File: content/io/cloudslang/cloud/openstack/utils/update_list.sl [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+#! @output error_message: error message if there was an error when executing, empty otherwise
+#! @result SUCCESS: successfully updated the specified list (return_code == '0')
+#! @result FAILURE: otherwise
+#!!#
+####################################################
+
+namespace: io.cloudslang.cloud.openstack.utils
+
+operation:
+  name: update_list
```
Comment (Bonczidai) [inline content/io/cloudslang/cloud/openstack/utils/update_list.sl]: can this be a flow which reuses `get_value` + `add_object_into_json_array` from json

**Comment by Bonczidai** [comment]
pretty nice - regarding json you should reuse the already existing content + enhance it under json

File: content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl (hunk 4 above)
Comment (mihaitusa) [inline content/io/cloudslang/cloud/openstack/servers/create_server_flow.sl]: @Bonczidai In "user_data" can be passed an Integer as well, but the user will receive an error if the value is not Base64 encoded. Furthermore if the value passed is not a valid script... will be ignored.

File: content/io/cloudslang/cloud/openstack/utils/add_json_object.sl (hunk 5 above)
Comment (mihaitusa) [inline content/io/cloudslang/cloud/openstack/utils/add_json_object.sl]: @Bonczidai  You're right! I will correct it.

File: content/io/cloudslang/cloud/openstack/utils/add_json_object.sl (hunk 6 above)
Comment (mihaitusa) [inline content/io/cloudslang/cloud/openstack/utils/add_json_object.sl]: @Bonczidai  I think that to.

File: content/io/cloudslang/cloud/openstack/utils/update_list.sl (hunk 7 above)
Comment (mihaitusa) [inline content/io/cloudslang/cloud/openstack/utils/update_list.sl]: @Bonczidai You right in here to. I will correct it.

File: content/io/cloudslang/cloud/openstack/utils/update_list.sl (hunk 8 above)
Comment (mihaitusa) [inline content/io/cloudslang/cloud/openstack/utils/update_list.sl]: @Bonczidai I didn't succeed to add a list under a certain key into a complex JSON structure.

**Pull request merged**